nomad/json/extensions.go

```diff
@@ -0,0 +1,30 @@
+package json
+
+import (
+	"reflect"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// init register all extensions used by the API HTTP server when encoding JSON
+func init() {
+	// TODO: this could be simplified by looking up the base type in the case of a pointer type
+	registerExtension(reflect.TypeOf(structs.Node{}), nodeExt)
+	registerExtension(reflect.TypeOf(&structs.Node{}), nodeExt)
+}
+
+// nodeExt adds the legacy field .Drain back to encoded Node objects
+func nodeExt(v interface{}) interface{} {
+	node := v.(*structs.Node)
+	if node == nil {
+		return nil
+	}
+	type NodeAlias structs.Node
```

Comment:
i'll add a comment about what's happening here and clarify the code.
the short answer is, i don't think so. what this is doing is creating a type definition (not an alias, bad naming aside) so that the embedded field has an [identical underlying field](https://golang.org/ref/spec#Conversions); this means that go gives us a free conversion function (for the two types, as well as pointers to them). and because it is a distinct type (unlike an alias), it doesn't recursively/infinitely trigger this encoding extension.

the choice of pointers was just for efficiency, to keep from having to copy the node fields. it's the difference between these two:
pointers: https://play.golang.org/p/WEMAk2YSYBs
no pointers: https://play.golang.org/p/D0oWKEt1lmd

long story short, it says "put a node in here, but pretend it's not a node". i couldn't find another way to do this.

i'll clean this bit up and try to explain the hack/magic.